src/base.mo

```diff
@@ -0,0 +1,448 @@
+      };
+      return result;
+    };
+
+    public func keyToIndex(bytes : [Nat8], pos : Nat16) : Nat64 {
+      let bit_pos = Nat8.fromNat16(pos & 7);
+      let ret = Nat8.toNat((bytes[Nat16.toNat(pos >> 3)] << bit_pos) >> bitshift);
+      return Nat64.fromIntWrap(ret);
+    };
+
+    public func find(nodes : Region, bytes : [Nat8]) : (Nat64, Nat64, Nat64, Nat16) {
+      var idx = keyToRootIndex(bytes);
+      var pos = root_bitlength;
+      var node : Nat64 = 0;
+      var old_leaf : Nat64 = 0;
+      loop {
+        switch (getChild(nodes, node, idx)) {
+          case (0) {
+            old_leaf := 0;
+            return (node, idx, old_leaf, pos);
+          };
+          case (n) {
+            if (n & 1 == 1) {
+              old_leaf := n;
+              return (node, idx, old_leaf, pos);
+            };
+            node := n;
+          };
+        };
+        idx := keyToIndex(bytes, pos);
+        pos +%= bitlength;
+      };
+      Debug.trap("Unreacheable");
+    };
+
+    public func put_(nodes : Region, leaves : Region, key : Blob) : ?(Bool, Nat64) {
+      assert key.size() == args.key_size;
+      let bytes = Blob.toArray(key);
+
+      let (node_, last_, old_leaf, pos_) = find(nodes, bytes);
+
+      var last = last_;
+      var node = node_;
+
+      if (old_leaf == 0) {
+        let ?leaf = newLeaf(leaves, key) else return null;
+
+        setChild(nodes, node, last, leaf);
+        return ?(true, (leaf >> 1));
+      };
+
+      let index = old_leaf >> 1;
+      let old_key = getKey(leaves, index);
+      if (key == old_key) {
+        return ?(false, index);
+      };
+
+      let old_bytes = Blob.toArray(old_key);
+      var pos = pos_;
+      label l loop {
+        let ?add = newInternalNode(nodes) else {
+          setChild(nodes, node, last, old_leaf);
+          return null;
+        };
+        setChild(nodes, node, last, add);
+        node := add;
+
+        let (a, b) = (keyToIndex(bytes, pos), keyToIndex(old_bytes, pos));
+        pos +%= bitlength;
+        if (a == b) {
+          last := a;
+        } else {
+          setChild(nodes, node, b, old_leaf);
+          let ?leaf = newLeaf(leaves, key) else return null;
+          setChild(nodes, node, a, leaf);
+          return ?(true, (leaf >> 1));
+        };
+      };
+      Debug.trap("Unreacheable");
+    };
+
+    public func lookup(key : Blob) : ?(Blob, Nat) {
+      assert key.size() == args.key_size;
+      let { leaves; nodes } = regions();
+
+      let bytes = Blob.toArray(key);
+
+      let (_, _, old_leaf, _) = find(nodes, bytes);
+      if (old_leaf == 0) return null;
+      let index = old_leaf >> 1;
+      return if (getKey(leaves, index) == key) ?(getValue(leaves, index), Nat64.toNat(index)) else null;
+    };
+
+    type Dir = { #forward; #reverse };
+
+    class Iterator(nodes : Region, dir : Dir) {
+      let stack = Array.init<(Nat64, Nat64)>(args.key_size * 8 / Nat16.toNat(bitlength), (0, 0));
```

Comment:
Suggesting here a line
`let forward : Bool = dir == #forward;`
and then use `if (forward)` in 3 places below as in the old code.